README.md
# ascent-2.4.3-LEGACY
Private emu Ascent 2.4.3 TBC legacy-wow

over 7200+ commits


ChangeLog.txt
15/04/12 15.00pm Author Bloodfangx | REV 1
- Start Working on ascent Branch

15/04/12 15.52pm Author Bloodfangx | REV 2
- Implement Sun++ Scripting 

15/04/12 16.00 Author Bloodfangx | REV 3
- Started converting Arc scripts from TbcE to Summit ascent format.

15/04/12 20.32 Author Bloodfangx | REV 4
- Update Base files with New Functions and Methods And Target Selections 
IE Target_SecondMostHated Etc.

15/04/12 20.33 Author Bloodfangx | REV 5
Scripts/Sunwell:Sucessfully Convert Sunwell Plat to Summit Ascent Format /Woooo

15/04/12 20.40 Author Bloodfangx | REV 6
- Implement More Target definitions into base

01/05/12 13.53 Author Bloodfangx | REV 7
- Expand Gm On Command to Bypass Instance requirements. | Affected files Level1.Cpp

01/05/12 14.07 Author Bloodfangx | REV 8
- Implement Custom Reset Health-Mana-Cooldown Core Script to duel system | Affected files Player.Cpp
- Update Dueling Boundary to 85 yards up from 75 | Affected Files Player.cpp
 
01/05/12 15.32 Author Bloodfangx | REV 9
- Implement Player update fields for dodge calculations
- Implement Correct Dodge calculation Methods

01/05/12 16.40 Author Bloodfangx | REV 10
- Implement Crit to dodge ratio

02/05/12 02.39 Author Bloodfangx | REV 11
- Try to fix Berserker Rage
- Fix Victory rush only usable once per kill.

02/05/12 03.22 Author Bloodfangx | REV 12
- Spell Fix Immunity effect for Berserker rage should work.

02/05/12 19.21 Author Bloodfangx | REV 13
- Implement HasVisialPosAurasOfNameHashWithCaster Into Unit 
- Add more Aura Checks Into Unit, MAX_POSITIVE_AURA_EXTENDED, START ETC.

02/05/12 19.32 Author Bloodfangx | REV 14
- Fix Vampiric Touch, will now restore mana on proc/use.

02/05/12 19.32 Author Bloodfangx | REV 15
- Attempt fixing Warlock health stones , warlock's should now recieve the item "healthstone" when using the specefic rank spell.

02/05/12 19.32 Author Bloodfangx | REV 16
- Attempt to fix Nether protection, procs of shadow and fire dmg DoT's ticks.

02/05/12 19.32 Author Bloodfangx | REV 17
- Implement SpellCastResult Into Spell

02/05/12 19.32 Author Bloodfangx | REV 18
- Implement Missing UNIT_FLAGS for npcs Thanks to arc for this.

09/05/12 15.32 Author Bloodfangx | REV 20
- Expand HasFreeSlots Function to include bg type if inside an arena.

09/05/12 19.32 Author Bloodfangx | REV 21
- Work On Updating Battleground system

09/05/12 19.32 Author Bloodfangx | REV 22
- Expand CanPlayerJoin 

09/05/12 19.32 Author Bloodfangx | REV 23
- Implement ProcOnNameHash 
- Port Spell Fixes From My Arcemu Branch.

09/05/12 19.32 Author Bloodfangx | REV 24
- Fix Energized
- Fix Spell focus trigger( Mystical Skyfire Diamond
- Fix Skyfire Swiftness ( Thundering Skyfire diamond)
- Fix Tsunami Talisman
- Fix Assassination Rogue 2 set bonus

09/05/12 19.32 Author Bloodfangx | REV 25
- Fix Rogue Sword Spec Rank 1-5

09/05/12 19.32 Author Bloodfangx | REV 26
- Fix Madness Of The Betrayer Proc Rate.


21/05/12 20.40 Author Bloodfangx | REV 27
- Implement IsNeutralFaction into unit.

21/05/12 20.40 Author Bloodfangx | REV 28
- Implement CanSee into unit.

21/05/12 21.22 Author Bloodfangx | REV 29
-Implement UnsafeCanOwnerAttackUnit into AIinterface

21/05/12 21.22 Author Bloodfangx | REV 30
- Scripts/Sunwell: Rewrote Sunblade Cabalist script using UpdateAi timers.

21/05/12 17.15 Author Bloodfangx | REV 31
- Implement Character enchant visuals to the character selection screen

21/05/12 18.20 Author Bloodfangx | REV 32
- Implement RepairItem2 call into Level2.cpp
- Implement Handlerepairitemscommand Into Level2
- Implement Repair Command


21/05/12 18.20 Author Bloodfangx | REV 33
- Implement Title defintions for Player known titles system
- Change Titles from Hex into Correct Id's enums.
- Implement PLAYER_KNOWN_TITLES


21/05/12 18.42 Author Bloodfangx | REV 34
- UpdateFields for 2.4.3 Thanks to arc for this.

21/05/12 18.42 Author Bloodfangx | REV 35
- Implement HasKnownTitle
- Implement SetKnownTitle. 

21/05/12 21.30 Author Bloodfangx | REV 36
- Add files for speeddetecting anti cheat.

27/05/12 05.03 Author Bloodfangx | REV 37
- Update Some Movement opcodes for 2.4.3


27/05/12 02.23 Author Bloodfangx | REV 38
- Implement Vapor trail Template AI for felmyst encounter


27/05/12 02.40 Author Bloodfangx | REV 39
- Update More of felmyst start implementing Encapsulate , need to hardcode a spell fix for this the gravity check isn't currently working.

27/05/12 02.51 Author Bloodfangx | REV 40
- Spell Fix Encapsulate add a fly effect on channel, DOESN'T WORK.

02/06/12 05.29 Author Bloodfangx | REV 41
- Implemented Gas nova on Felmyst
- Add in Phase_Flight check into AIUpdate for felmyst

02/06/12 05.40 Author Bloodfangx | REV 42
- Add UNIT_FIELD_BOUNDINGRADIUS and UNIT_FIELD_COMBATREACH checks to felmyst and Brut, some reason it isn't working as it should tho :|.
- Add Init to OnCombatStop for Felmyst, should remove the problem with spell timers activating to fast as combat started.

02/06/12 05.40 Author Bloodfangx | REV 43
Scripts/Sunwell/Twins
- update todo
- Remove Old Static casts on Twins encounter in SWP and use new TO_PLAYER And unit checks.

03/06/12 05.40 Author Bloodfangx | REV 44
Scripts/Sunwell/Felmyst
- Implement Mob_felmust_vaporai
- Add init() to class constructor will fix the spell timers initiatingto fast


03/06/12 05.58 Author Bloodfangx | REV 45
Scripts/Sunwell/Felmyst
- Implement Mob_felmyst_vaporAI

17/06/12 05.48 Author Bloodfangx | REV 46
Script/Base
- implemented Event_OnCombatStop - it is used with AddEmote event part by using this, text emote will show when Npc goes back to home position (reset)

23/06/12 04.45 Author Bloodfangx | REV 47
Script/WorldPvp
- implemented Zangamarsh pvp and Terrokar Forest Pvp.


src/ascent-world/Quest.h
/*
 * Ascent MMORPG Server
 * Copyright (C) 2005-2008 Ascent Team <http://www.ascentemu.com/>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#ifndef WOWSERVER_QUEST_H
#define WOWSERVER_QUEST_H

using namespace std;
/*
2.3.0 research
not available because low level = 1
available but quest low level = 2
question mark = 3 (not got objectives)
blue question mark = 4
blue exclamation mark = 5
yellow exclamation mark = 6
yellow question mark = 7
finished = 8
132 error = 9
*/

enum QUEST_STATUS
{
	QMGR_QUEST_NOT_AVAILABLE			= 0x00,	// There aren't quests avaiable.				   | "No Mark"
	QMGR_QUEST_AVAILABLELOW_LEVEL	   = 0x01,	// Quest avaiable, and your level isnt enough.	 | "Gray Quotation Mark !"
	QMGR_QUEST_CHAT					 = 0x02,	// Quest avaiable it shows a talk baloon.		  | "No Mark"
	QMGR_QUEST_NOT_FINISHED			 = 0x03,	// Quest isnt finished yet.						| "Gray Question ? Mark"
	QMGR_QUEST_REPEATABLE_FINISHED		= 0x04,
	QMGR_QUEST_REPEATABLE			   = 0x05,	// Quest repeatable								| "Blue Question ? Mark" 
	QMGR_QUEST_AVAILABLE				= 0x06,	// Quest avaiable, and your level is enough		| "Yellow Quotation ! Mark" 
	QMGR_QUEST_FINISHED				 = 0x08,	// Quest has been finished.						| "Yellow Question  ? Mark" (7 has no minimap icon)
	//QUEST_ITEM_UPDATE				 = 0x06	 // Yellow Question "?" Mark. //Unknown
};

enum QUESTGIVER_QUEST_TYPE
{
	QUESTGIVER_QUEST_START  = 0x01,
	QUESTGIVER_QUEST_END	= 0x02,
};

enum QUEST_TYPE
{
	QUEST_GATHER	= 0x01,
	QUEST_SLAY	  = 0x02,
};

enum QUEST_FLAG
{
	QUEST_FLAG_NONE		  = 0,   
	QUEST_FLAG_DELIVER	   = 1,   
	QUEST_FLAG_KILL		  = 2,   
	QUEST_FLAG_SPEAKTO	   = 4,
	QUEST_FLAG_REPEATABLE	= 8,   
	QUEST_FLAG_EXPLORATION   = 16,
	QUEST_FLAG_TIMED		 = 32,
	QUEST_FLAG_REPUTATION	= 128,
};

enum FAILED_REASON
{
	FAILED_REASON_FAILED			= 0,
	FAILED_REASON_INV_FULL			= 4,
    FAILED_REASON_DUPE_ITEM_FOUND   = 17,
};

enum INVALID_REASON
{
	INVALID_REASON_DONT_HAVE_REQ			= 0,
	INVALID_REASON_DONT_HAVE_LEVEL			= 1,
	INVALID_REASON_DONT_HAVE_RACE			= 6,
	INVALID_REASON_COMPLETED_QUEST			= 7,
	INVALID_REASON_HAVE_TIMED_QUEST			= 12,
	INVALID_REASON_HAVE_QUEST				= 13,
//	INVALID_REASON_DONT_HAVE_REQ_ITEMS	  = 0x13,
//	INVALID_REASON_DONT_HAVE_REQ_MONEY	  = 0x15,
	INVALID_REASON_DONT_HAVE_EXP_ACCOUNT	= 16,
	INVALID_REASON_DONT_HAVE_REQ_ITEMS		= 21, //changed for 2.1.3
	INVALID_REASON_DONT_HAVE_REQ_MONEY		= 23,
	INVALID_REASON_UNKNOW26					= 26, //"you have completed 10 daily quests today"
	INVALID_REASON_UNKNOW27					= 27,//"You cannot completed quests once you have reached tired time"
};

enum QUEST_SHARE
{
	QUEST_SHARE_MSG_SHARING_QUEST   = 0,
	QUEST_SHARE_MSG_CANT_TAKE_QUEST = 1,
	QUEST_SHARE_MSG_ACCEPT_QUEST	= 2,
	QUEST_SHARE_MSG_REFUSE_QUEST	= 3,
	QUEST_SHARE_MSG_TOO_FAR		 = 4,
	QUEST_SHARE_MSG_BUSY			= 5,
	QUEST_SHARE_MSG_LOG_FULL		= 6,
	QUEST_SHARE_MSG_HAVE_QUEST	  = 7,
	QUEST_SHARE_MSG_FINISH_QUEST	= 8,
};

class QuestScript;
#pragma pack(push,1)
struct Quest
{
	uint32 id;
	uint32 zone_id;
	uint32 quest_sort;
	uint32 quest_flags;
	uint32 min_level;
	uint32 max_level;
	uint32 type;
	uint32 required_races;
	uint32 required_class;
	uint32 required_tradeskill;
	uint32 required_tradeskill_value;
	uint32 required_rep_faction;
	uint32 required_rep_value;

	uint32 time;
	uint32 special_flags;
	
	uint32 previous_quest_id;
	uint32 next_quest_id;

	uint32 srcitem;
	uint32 srcitemcount;

	char * title;
	char * details;
	char * objectives;
	char * completiontext;
	char * incompletetext;
	char * endtext;

	char * objectivetexts[4];

	uint32 required_item[4];
	uint32 required_itemcount[4];

	uint32 required_mob[4];
	uint32 required_mobcount[4];
	uint32 required_spell[4];

	uint32 reward_choiceitem[6];
	uint32 reward_choiceitemcount[6];

	uint32 reward_item[4];
	uint32 reward_itemcount[4];

	uint32 reward_repfaction[2];
	int32 reward_repvalue[2];
	uint32 reward_replimit;

	uint32 reward_rank;

	uint32 reward_money;
	uint32 reward_xp;
	uint32 reward_spell;
	uint32 effect_on_player;
	
	uint32 point_mapid;
	uint32 point_x;
	uint32 point_y;
	uint32 point_opt;

	uint32 required_money;
	uint32 required_triggers[4];
	uint32 required_quests[4];
	uint32 receive_items[4];
	uint32 receive_itemcount[4];
	int is_repeatable;

	uint32 count_required_mob;
	uint32 count_requiredquests;
	uint32 count_requiredtriggers;
	uint32 count_receiveitems;
	uint32 count_reward_choiceitem;
	uint32 count_required_item;
	uint32 required_mobtype[4];
	uint32 count_reward_item;
	uint32 reward_xp_as_money;

	QuestScript* pQuestScript;
};
#pragma pack(pop)

enum QUEST_MOB_TYPES
{
	QUEST_MOB_TYPE_CREATURE = 1,
	QUEST_MOB_TYPE_GAMEOBJECT = 2,
};

class QuestScript;
#define CALL_QUESTSCRIPT_EVENT(obj, func) if(static_cast<QuestLogEntry*>(obj)->GetQuest()->pQuestScript != NULL) static_cast<QuestLogEntry*>(obj)->GetQuest()->pQuestScript->func

class SERVER_DECL QuestLogEntry : public EventableObject
{
	friend class QuestMgr;

public:

	QuestLogEntry();
	~QuestLogEntry();

	ASCENT_INLINE Quest* GetQuest() { return m_quest; };
	void Init(Quest* quest, Player* plr, uint32 slot);

	bool CanBeFinished();
	void SubtractTime(uint32 value);
	void SaveToDB(QueryBuffer * buf);
	bool LoadFromDB(Field *fields);
	void UpdatePlayerFields();

	void SetTrigger(uint32 i);
	void SetMobCount(uint32 i, uint32 count);

	bool IsUnitAffected(Unit* target);
	ASCENT_INLINE bool IsCastQuest() { return iscastquest;}
	void AddAffectedUnit(Unit* target);
	void ClearAffectedUnits();

	void SetSlot(int32 i);
	void Finish();

	void SendQuestComplete();
	void SendUpdateAddKill(uint32 i);
	ASCENT_INLINE uint32 GetMobCount(uint32 i) { return m_mobcount[i]; }
	ASCENT_INLINE uint32 GetExploredAreas(uint32 i) { return m_explored_areas[i]; }

	ASCENT_INLINE uint32 GetBaseField(uint32 slot)
	{
		return PLAYER_QUEST_LOG_1_1 + (slot * 4);
	}

private:
	uint32 completed;

	bool mInitialized;
	bool mDirty;

	Quest *m_quest;
	Player *m_plr;
	
	uint32 m_mobcount[4];
	uint32 m_explored_areas[4];

	std::set<uint64> m_affected_units;
	bool iscastquest;

	uint32 m_time_left;
	int32 m_slot;
};

#endif
